ctapipe 0.19 & co.

aleberti opened the pull request:
Starting from the branch that Elisa created to get independence from `lstchain`. I would use this branch to get everything done to update to `ctapipe` v0.19 and corresponding `lstchain`.

aleberti commented:
FYI, the current code can already read LST-1 DL1 files created with lstchain v0.10. Not sure about MC files, I think there is no MC production yet reprocessed with v0.10.

So, for the moment, it may be possible to keep the usual format we used for the analysis until now.

In any case, we (I) cannot switch to the standard ctapipe DL1 format for MAGIC data until https://github.com/cta-observatory/ctapipe/issues/2015 is sorted out.

aleberti commented:
Ok, until now I:

- updated the environment to ctapipe 0.19.2 and lstchain 0.10.x (0.10.3 at the moment)
- reintroudced the lstchain dependencies and created a specific module called, with a lot of fantasy from my side, `lstchain`, containing the functions that Elisa ported from `lstchain`. I put them in a specific module so that it will be easier in the future to update and import those functions if we will need them
- the only "reference" to lstchain are the functions `create_hdu_index_hdu` and `create_obs_index_hdu`, now in `magicctapipe.lstchain.high_level` module. I kept them because in that way we can have the header saying that the creator of the index files was `magicctapipe` and not `lstchain`. A small thing, but I wanted to have it

So now we can start from a semi-clean state and see what other changes are needed with the new versions of ctapipe and lstchain. Unfortunately I think that on the LST side, after we merge #141 , tests will fail until we replace the test files (MC and LST real data) with those reprocessed with the latest lstchain (and therefore ctapipe) versions.

Elisa-Visentin commented:
As for tests (and data analysis)... Real LST data could make the code crash (I don't know if they already reprocessed every LST run with lstchain 0.10; the test file can be updated just before merging this PR and tests should not fail; in the meanwhile, the magic-only tests should not fail); as for MCs, DL0 should be processed by these scripts (I used them a few months ago), but the MCs already processed (DL1-->IRF) will probably fail because of the different SubArray version (LST). But MC reprocessing will be needed by Raniere's PR too (in particular to use the scripts that automatically match data-MCs), so we can do this after we merge these PRs (and any other needed and 'urgent' major change affecting data format), before starting massive shared data analysis

jsitarek commented:
Hi, I'm starting to run this branch for test MCs. 
At the mc_dl0_to_dl1 step I got this error (which does not stop the execution):
Column tels_with_trigger with value None of type <class 'NoneType'>  of container SimEventInfoContainer in table parameters not writable, skipping
I guess this field will be used at least to mark the MAGIC stereo triggers.

Elisa-Visentin commented:
this is not for the stereo triggers (there is a MAGIC-stereo parameter) but
for the telescope combination (i.e., M1_L1...). Maybe I know the reason.
I'm doing the bug fix PR and then I will try to fix this

Il giorno ven 15 dic 2023 alle ore 18:53 Julian Sitarek <
***@***.***> ha scritto:

> Hi, I'm starting to run this branch for test MCs.
> At the mc_dl0_to_dl1 step I got this error (which does not stop the
> execution):
> Column tels_with_trigger with value None of type <class 'NoneType'> of
> container SimEventInfoContainer in table parameters not writable, skipping
> I guess this field will be used at least to mark the MAGIC stereo triggers.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cta-observatory/magic-cta-pipe/pull/148#issuecomment-1858265594>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A43O4NHFLS7VYD5SUQURXM3YJSFBHAVCNFSM6AAAAAA4BQWTCCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNJYGI3DKNJZGQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

-- 
------------------------


Indirizzo istituzionale di posta elettronica 
degli studenti e dei laureati dell'Università di TorinoOfficial University 
of Turin email address for students and graduates 

Elisa-Visentin commented:
Ok, can you try it again?
Thanks


Il giorno ven 15 dic 2023 alle ore 20:12 Elisa Visentin <
***@***.***> ha scritto:

> this is not for the stereo triggers (there is a MAGIC-stereo parameter)
> but for the telescope combination (i.e., M1_L1...). Maybe I know the
> reason. I'm doing the bug fix PR and then I will try to fix this
>
> Il giorno ven 15 dic 2023 alle ore 18:53 Julian Sitarek <
> ***@***.***> ha scritto:
>
>> Hi, I'm starting to run this branch for test MCs.
>> At the mc_dl0_to_dl1 step I got this error (which does not stop the
>> execution):
>> Column tels_with_trigger with value None of type <class 'NoneType'> of
>> container SimEventInfoContainer in table parameters not writable, skipping
>> I guess this field will be used at least to mark the MAGIC stereo
>> triggers.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/cta-observatory/magic-cta-pipe/pull/148#issuecomment-1858265594>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/A43O4NHFLS7VYD5SUQURXM3YJSFBHAVCNFSM6AAAAAA4BQWTCCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNJYGI3DKNJZGQ>
>> .
>> You are receiving this because you commented.Message ID:
>> ***@***.***>
>>
>

-- 
------------------------


Indirizzo istituzionale di posta elettronica 
degli studenti e dei laureati dell'Università di TorinoOfficial University 
of Turin email address for students and graduates 

jsitarek commented:
Hi @Elisa-Visentin ,

now it completely breaks: 

--> 1 event (event ID: 1102, telescope 1) could not survive the image cleaning. Skipping...
--> 7 event (event ID: 2801, telescope 1) could not survive the image cleaning. Skipping...
Error writing col tels_with_trigger of container SimEventInfoContainer
Traceback (most recent call last):
  File "/home/julian.sitarek/prog/mcp_ctapipe019/magicctapipe/scripts/lst1_magic/lst1_magic_mc_dl0_to_dl1.py", line 480, in <module>
    main()
  File "/home/julian.sitarek/prog/mcp_ctapipe019/magicctapipe/scripts/lst1_magic/lst1_magic_mc_dl0_to_dl1.py", line 471, in main
    mc_dl0_to_dl1(args.input_file, args.output_dir, config, args.focal_length_choice)
  File "/home/julian.sitarek/prog/mcp_ctapipe019/magicctapipe/scripts/lst1_magic/lst1_magic_mc_dl0_to_dl1.py", line 371, in mc_dl0_to_dl1
    writer.write(
  File "/home/julian.sitarek/anaconda3/envs/magic-lst_0_19/lib/python3.11/site-packages/ctapipe/io/hdf5tableio.py", line 478, in write
    self._append_row(table_name, containers)
  File "/home/julian.sitarek/anaconda3/envs/magic-lst_0_19/lib/python3.11/site-packages/ctapipe/io/hdf5tableio.py", line 448, in _append_row
    row[colname] = value
    ~~~^^^^^^^^^
  File "tables/tableextension.pyx", line 1679, in tables.tableextension.Row.__setitem__
ValueError: could not broadcast input array from shape (3,) into shape (2,)

in different files I get different combinations:
ValueError: could not broadcast input array from shape (3,) into shape (1,)
ValueError: could not broadcast input array from shape (2,) into shape (3,)

I think  the size of the table varies from one event to another which makes a problem when it is being written.

aleberti commented:
One has to use a solution similar to what is implemented in `magic_calib_to_dl1.py`, indeed `tels_with_trigger` is a list with variable length, but HDF5Writer does not support it. In `magic_calib_to_dl1.py` it uses `TEL_COMBINATIONS` to give the corresponding combination type (I guess now this would need to be changed because of the issue that was found by Eisa):

```python
TEL_COMBINATIONS = {
    "LST1_M1": [1, 2],  # combo_type = 0
    "LST1_M1_M2": [1, 2, 3],  # combo_type = 1
    "LST1_M2": [1, 3],  # combo_type = 2
    "M1_M2": [2, 3],  # combo_type = 3
}  
```

Anyway, `tels_with_trigger` is not really used to select events at the moment.

jsitarek commented:
I'm running the pipeline on some Crab samples.
Most went fine, but one of the runs: 20201214_M2_05093664 fails, while it was working in the original analysis:

Input file: /home/julian.sitarek/ws/analiza/Crab_2021/an18_ctapipe019/data/dl1_MAGIC/../MARS//20201214_M2_05093664.001_Y_CrabNebula-W0.40+035.root

    prescaler = list(prescaler_array[0])
                     ~~~~~~~~~~~~~~~^^^
IndexError: index 0 is out of bounds for axis 0 with size 0

I am not sure why it does not work since Trigger tree has 9 entries for this subrun.

aleberti commented:
> I'm running the pipeline on some Crab samples. Most went fine, but one of the runs: 20201214_M2_05093664 fails, while it was working in the original analysis:
> 
> Input file: /home/julian.sitarek/ws/analiza/Crab_2021/an18_ctapipe019/data/dl1_MAGIC/../MARS//20201214_M2_05093664.001_Y_CrabNebula-W0.40+035.root
> 
> ```
> prescaler = list(prescaler_array[0])
>                  ~~~~~~~~~~~~~~~^^^
> ```
> 
> IndexError: index 0 is out of bounds for axis 0 with size 0
> 
> I am not sure why it does not work since Trigger tree has 9 entries for this subrun. Either way this is something to be fixed in ctapipe_io_magic

the problem is not that file, but one of the subruns (since I assume you are running with `process_run` set):

```
In [5]: for filef in file_list:
   ...:     rootf = uproot.open(filef)
   ...:     print(f"{filef}: {rootf['Trigger']['MTriggerPrescFact.fPrescFact'].array(library='np').size}")
   ...: 
20201214_M2_05093664.001_Y_CrabNebula-W0.40+035.root: 9
20201214_M2_05093664.002_Y_CrabNebula-W0.40+035.root: 9
20201214_M2_05093664.003_Y_CrabNebula-W0.40+035.root: 8
20201214_M2_05093664.004_Y_CrabNebula-W0.40+035.root: 9
20201214_M2_05093664.005_Y_CrabNebula-W0.40+035.root: 9
20201214_M2_05093664.006_Y_CrabNebula-W0.40+035.root: 8
20201214_M2_05093664.007_Y_CrabNebula-W0.40+035.root: 9
20201214_M2_05093664.008_Y_CrabNebula-W0.40+035.root: 9
20201214_M2_05093664.009_Y_CrabNebula-W0.40+035.root: 8
20201214_M2_05093664.010_Y_CrabNebula-W0.40+035.root: 9
20201214_M2_05093664.011_Y_CrabNebula-W0.40+035.root: 9
20201214_M2_05093664.012_Y_CrabNebula-W0.40+035.root: 8
20201214_M2_05093664.013_Y_CrabNebula-W0.40+035.root: 8
20201214_M2_05093664.014_Y_CrabNebula-W0.40+035.root: 2
20201214_M2_05093664.015_Y_CrabNebula-W0.40+035.root: 0
20201214_M2_05093664.016_Y_CrabNebula-W0.40+035.root: 51
```

so the culprit is actually 20201214_M2_05093664.015_Y_CrabNebula-W0.40+035.root.

aleberti commented:
@jsitarek should I fix those files for you?

aleberti commented:
Hi again. I cannot fix 20201214_M2_05093664.015_Y_CrabNebula-W0.40+035.root. In the report file, there is no report (for any subsystem) between 2020-12-14 04:48:25 and 2020-12-14 04:50:46 and this subrun starts at 14.12.2020 04:49:52.000 and stops at 14.12.2020 04:50:46.000. For the last run, for some reason its duration is 10min, so it is normal that the entries are so many.

jsitarek commented:
> Hi again. I cannot fix 20201214_M2_05093664.015_Y_CrabNebula-W0.40+035.root. In the report file, there is no report (for any subsystem) between 2020-12-14 04:48:25 and 2020-12-14 04:50:46 and this subrun starts at 14.12.2020 04:49:52.000 and stops at 14.12.2020 04:50:46.000. For the last run, for some reason its duration is 10min, so it is normal that the entries are so many.
> 
> I would exclude them, it should not change much.

Thanks a lot! When I remove those two last subruns conversion to DL1 files starts without problems. I didn't realize that the problem was with later files because I got a crash at the very beginning of the processing and it only mentioned the (first) subrun that I gave as the input. It would be good to introduce a check that skips automatically such broken subruns. I will make a new issue in MCP on this.

Elisa-Visentin commented:
Now tests are of up to IRF (included). I also tried to 'fix' the tels_with_trigger issue (by converting the list of IDs into a single int, that can be written by ctapipe writer).
Still issue with interpolation (Dl2 to Dl3) due to some changes in pyirf, but I will try to solve them before the School

jsitarek commented:
Hi @Elisa-Visentin 
I made a small commit (https://github.com/cta-observatory/magic-cta-pipe/pull/148/commits/d4cbcb9cdc6c4c38d1b1f72677e715ae08beec3b) with changes needed to make DL3 files with interpolation:

a quick test with Crab sample gives reasonable spectrum:

![crab_sed_logAeff](https://github.com/cta-observatory/magic-cta-pipe/assets/33022433/3b23599a-faf8-432c-b0d0-e47c129dd932)

Elisa-Visentin commented:
Thank you Julian, I will test it on my data, too (and send you the
results). I will also put again the DL3 tests in the pipeline and run them.

Il mar 2 gen 2024, 18:28 Julian Sitarek ***@***.***> ha
scritto:

> Hi @Elisa-Visentin <https://github.com/Elisa-Visentin>
> I made a small commit (d4cbcb9
> <https://github.com/cta-observatory/magic-cta-pipe/commit/d4cbcb9cdc6c4c38d1b1f72677e715ae08beec3b>)
> with changes needed to make DL3 files with interpolation:
>
> a quick test with Crab sample gives reasonable spectrum:
>
> crab_sed_logAeff.png (view on web)
> <https://github.com/cta-observatory/magic-cta-pipe/assets/33022433/3b23599a-faf8-432c-b0d0-e47c129dd932>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cta-observatory/magic-cta-pipe/pull/148#issuecomment-1874325593>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A43O4NEP35VMUAYJMORASM3YMQ7SFAVCNFSM6AAAAAA4BQWTCCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNZUGMZDKNJZGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

-- 
------------------------


Indirizzo istituzionale di posta elettronica 
degli studenti e dei laureati dell'Università di TorinoOfficial University 
of Turin email address for students and graduates 

jsitarek commented on magicctapipe/io/io.py:
```diff
@@ -990,17 +988,19 @@ def load_mc_dl2_data_file(input_file, quality_cuts, event_type, weight_type_dl2)
+        # viewcone = 0 * u.deg
+        min_viewcone_radius = max_viewcone_radius
+    # else:
+    # viewcone = max_viewcone_radius
```
the commented out parts about the viewcone can be probably removed

aleberti replied:
Done.

jsitarek commented on magicctapipe/io/io.py:
```diff
@@ -1075,6 +1075,14 @@ def load_dl2_data_file(input_file, quality_cuts, event_type, weight_type_dl2):
+
+    logger.info(
+        f"{event_before_nan_filter - event_after_nan_filter} "
+        f"stereo events removed because of NaN values."
```
for the Crab files that I was using for the tests I did not get a single event rejected like this, but I guess it must have happened to you that's why you introduce this check

aleberti replied:
indeed, it was another sample, but it gave problems because some events had NaN for the reconstructed alt/az and ra/dec, and then it gave errors when multiplying by a unit.

jsitarek commented on magicctapipe/lstchain/__init__.py:
do we need this empty file? it will make a package out of lstchain directory, but we are also using lstchain 0.10 directly

aleberti replied:
I added this lstchain module because at the beginning, this branch was created by Elisa copying the lstchain functions within other submodules of MCP, because the new version of lstchain based on ctapipe 0.19 was not out yet. Then the new lstchain got out, so I decided to move all the lstchain functions we use in this "lstchain" submodule. However, we do not use them at the moment, we use the ones from lstchain directly. For me it is ok to remove all the submodule. It just means that if for some reason we find ourselves in the same situation, we just wait for the new lstchain version to come out and that's it. If you agree, I will remove the whole submodule (it was also creating issues for the docs generation actually).

jsitarek replied:
ok, thx for the explanation. I think the copied functions are used only in one place in the code: create_dl3_index_files.py
the copied functions have a slightly different API than the ones in lstchain, which is very easy to fix, but also the header of the output run list fits file is filled differently: in MCP version magicctapipe with its version is given as the creator, while if we use lstchain functions it will specify lstchain with its version. Since this is done within a variable that is not exported it is not easy to change it if we use the lstchain function. However since those are just the index files and the bulk of the function to create this index files is either way from lstchain I think it is fine to have "lstchain" written in the header of the file. 
So I would be for the solution of removing all the lstchain-copied files. Now they jsut make it confusing in the code which version is really used, and if we need to do the same procedure in the future we would either way need to copy the never version of these functions.

aleberti replied:
Ok, done.

jsitarek commented on magicctapipe/lstchain/image/cleaning.py:
```diff
@@ -0,0 +1,38 @@
+__all__ = ["apply_dynamic_cleaning"]
+
+
+def apply_dynamic_cleaning(image, signal_pixels, threshold, fraction):
```
this function is exactly the same as in lstchain 0.10, cant we use it from there?

aleberti replied:
See comment on the lstchain __init__ file.

jsitarek commented on magicctapipe/lstchain/image/modifier.py:
```diff
@@ -0,0 +1,138 @@
+SMEAR_PROBABILITIES = np.full(N_PIXEL_NEIGHBORS, 1 / N_PIXEL_NEIGHBORS)
+
+
+def add_noise_in_pixels(
```
those functions also look the same (with some cosmetic differences) like in lstchain 0.10.6

aleberti replied:
See comment on the lstchain __init__ file

jsitarek commented on magicctapipe/lstchain/reco/utils.py:
```diff
@@ -0,0 +1,23 @@
+import numpy as np
+
+
+def add_delta_t_key(events):
```
again, same function as in lstchain 0.10

aleberti replied:
See comment on the lstchain __init__ file

jsitarek commented on magicctapipe/scripts/lst1_magic/lst1_magic_mc_dl0_to_dl1.py:
```diff
@@ -335,6 +344,18 @@ def mc_dl0_to_dl1(input_file, output_dir, config, focal_length):
+                tels_trigger = ""
+                for i in tels_with_trigger:
+                    tels_trigger = tels_trigger + str(i) + "000"
+                tels_trigger.rstrip("000")
```
rstrip ("000") does the same as rstrip("0") - will remove all 0's from the end of the string. As long as the tel_id are <10 it is fine, but if you have ...10000 it will turn it into ...1 rather than ...10

jsitarek commented on magicctapipe/scripts/lst1_magic/lst1_magic_mc_dl0_to_dl1.py:
```diff
@@ -352,6 +373,7 @@ def mc_dl0_to_dl1(input_file, output_dir, config, focal_length):
                     n_pixels=n_pixels,
                     n_islands=n_islands,
                     magic_stereo=magic_stereo,
+                    tels_with_trigger=float(tels_trigger),
```
saving it as a float is a bad idea. It already fails with 6 telescopes (this is in ipython, which seem to save still many digits for a float, but saving this to the file I think it will fail even faster):
```
In [7]: a="1000200030004005006"                                                 

In [8]: float(a)                                                                
Out[8]: 1.000200030004005e+18

In [9]: int(float(a))                                                           
Out[9]: 1000200030004004992
```

one could save it as int, but I'm afraid that this can also run out of int size when writing to file. Why not simply save it as a string separating the tel_ids with a coma?

Elisa-Visentin replied:
When trying to save it as a string it complained about this and did not save it. But I can try again

Elisa-Visentin replied:
Ok, if we use a string, the stereo step crashes:

```
Reconstructing the stereo parameters...
0 events
10 events
Traceback (most recent call last):
  File "/home/evisentin/mambaforge/envs/ctap19/bin/lst1_magic_stereo_reco", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/evisentin/mambaforge/envs/ctap19/lib/python3.11/site-packages/magicctapipe/scripts/lst1_magic/lst1_magic_stereo_reco.py", line 441, in main
    stereo_reconstruction(args.input_file, args.output_dir, config, args.magic_only)
  File "/home/evisentin/mambaforge/envs/ctap19/lib/python3.11/site-packages/magicctapipe/scripts/lst1_magic/lst1_magic_stereo_reco.py", line 370, in stereo_reconstruction
    save_pandas_data_in_table(
  File "/home/evisentin/mambaforge/envs/ctap19/lib/python3.11/site-packages/magicctapipe/io/io.py", line 1382, in save_pandas_data_in_table
    f_out.create_table(group_name, table_name, createparents=True, obj=data_array)
  File "/home/evisentin/mambaforge/envs/ctap19/lib/python3.11/site-packages/tables/file.py", line 1002, in create_table
    descr, _ = descr_from_dtype(obj.dtype, ptparams=self.params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/evisentin/mambaforge/envs/ctap19/lib/python3.11/site-packages/tables/description.py", line 846, in descr_from_dtype
    raise NotImplementedError(
NotImplementedError: structured arrays with columns with type description ``object`` are not supported yet, sorry
```
If we try to write it as a list, it won't be written.
Any idea? Do we really need this parameter?

Elisa-Visentin replied:
what about writing it as an int which refer to each possible telescope combination (i.e., as the combo_type)?

aleberti replied:
strings are problematic because there is no fixed size. In principle they can be written out only if a maximum number of characters is specified. But the solution with the binary encoding may be suitable (i.e. the same as writing out a int value).

jsitarek commented on magicctapipe/scripts/lst1_magic/magic_calib_to_dl1.py:
```diff
@@ -162,6 +167,38 @@ def magic_calib_to_dl1(input_file, output_dir, config, max_events, process_run=F
+    # Reset the telescope IDs of the subarray description
+    if event_source.is_hast:
+        tel_positions_magic_lst = {
+            1: [-8.09, 77.13, 0.78] * u.m,  # LST-1
```
but what happens if somebody uses different tel_ids in config file for HaST data?
maybe we can at least introduce a check that for HaST data lst-1 is 1  etc. ?

jsitarek commented on magicctapipe/scripts/lst1_magic/magic_calib_to_dl1.py:
```diff
@@ -162,6 +167,38 @@ def magic_calib_to_dl1(input_file, output_dir, config, max_events, process_run=F
+        subarray_magic = SubarrayDescription(
+            "MAGIC-LST-Array", tel_positions_magic_lst, tel_descriptions_magic_lst
+        )
+    else:
```
so if you have joint data but with a software trigger it will also only use MAGIC positions ?? something I do not understand here because the new code works with the 2019/20 data that are pre-HaST

aleberti replied:
this was not changed wrt before. When we did not have HaST, only the MAGIC positions were written out in the subarray, with shift of index (1-->2, 2-->3, so that LST-1 was 1). Then the subarray information is changed again in the coincidence scripts (from line 562).

jsitarek replied:
ok, let's keep it like this for this PR, but I think we will need to revise it in the future, because the logic of the code is confusing

jsitarek commented on magicctapipe/scripts/lst1_magic/magic_calib_to_dl1.py:
```diff
@@ -308,6 +345,25 @@ def magic_calib_to_dl1(input_file, output_dir, config, max_events, process_run=F
             elif tel_id == 2:
                 event_info.tel_id = config["mc_tel_ids"]["MAGIC-II"]  # MAGIC-II
 
+            if event.trigger.tels_with_trigger == [1, 2]:
```
this conversion here from the old numbering to the new numbering is really confusing. It needs at least some explanations.

aleberti replied:
It is not a new numbering. Anyway, if one uses 4LSTs, this will not work (realized just now, because I was focused in the 3 real telescopes case only). As we discussed via email, also here we can use the binary format stored as int, so that it will be unique for any telescope combination.

jsitarek replied:
ok, again let's merge it as it is, but we can simplify the code in another PR (just a single loop over tels_with_trigger and converting them into assigned_tel_ids)

jsitarek commented on environment.yml:
```diff
@@ -3,37 +3,45 @@ name: magic-lst
-      - ctaplot~=0.5.5
-      - pyirf~=0.6.0
+  - numba>=0.56
+  - uproot
```
setup requires uproot ~5.0, maybe it can be already requested in the same version in the environment file?

aleberti replied:
can be done, anyway I do not expect an uproot 6.x soon :)

jsitarek commented on environment.yml:
```diff
@@ -3,37 +3,45 @@ name: magic-lst
+  - numba>=0.56
+  - uproot
+  - pymongo
+  - jinja2 # pinned for bokeh 1.0 compatibility
```
similar here, setup uses ~3.0.2, while here there is no version specified, despite the comment says that it is "pinned"

Elisa-Visentin replied:
I think we should not use 'fixed' versions unless necessary to create a working environment because versions depends on the versions of the other dependencies and we could run into clashes when we update the environment/when new versions are released. The comment is old, we should check again the compatibility (btw, conda search for compatible versions, if not fixed, when creating the environment).  @aleberti: what about this?

aleberti replied:
I think it is ok to have jinja~=3.0.2, because in any case it is the one ctapipe has, so we need to be compatible in this sense.

jsitarek commented on setup.py:
this one is not required any more?

aleberti replied:
it is not required anymore, replaced by setup.cfg

jsitarek reviewed (changes requested):
I put some comments. Most are questions or suggestions, but the part about saving telescopes with trigger can lead to problems if kept as it is now.

aleberti commented:
@jsitarek when/if you have time, I implemented the check of the subarray for MAGIC DL1 files. I also implemented the encoding of the tels_with_trigger as an int (i.e. sum of 2^tel_id for the tel_ids of telescopes that triggered a specific event) as you suggested. In principle, this should close the comments you left. Let us know!

jsitarek commented:
> @jsitarek when/if you have time, I implemented the check of the subarray for MAGIC DL1 files. I also implemented the encoding of the tels_with_trigger as an int (i.e. sum of 2^tel_id for the tel_ids of telescopes that triggered a specific event) as you suggested. In principle, this should close the comments you left. Let us know!

Sorry for slow feedback on this. I checked all the pending issues. There were some that involved pathological cases with people playing with tel_id numbering, but this is not related to ctapipe0.19 compatibility so we can look into this later on with a new PR. Concerning this test that you implemented, I only saw the one checking consistency between different subarray information if multiple files are read at once, while we were discussing that the telescope numbering should be set at the first analysis step and then it should not be changed by new settings in input card, or there should be a check that the settings that are used in the input card do not clash with what is read from the subarray, but again, I think we can do it properly in another PR. Let's merge this one :-).

aleberti commented:
@jsitarek ok, I misunderstood for the latter point. I think it must be implemented indeed, but it may take some time because it would imply at some point a check between the settings used in MC and real data (I imagine a case where people use for some reason a separate input card for MC and real data, but anyway they should be in any case consistent). Let's shortly discuss tomorrow, then we can merge. We can open specific issues so that we do not forget about this.
